Implement Get ETH page

wackerow opened the pull request:
## Description
- Implements Get ETH page as `/src/pages/get-eth.tsx`
- Implements `CentralizedExchanges` component, and associated `useCentralizedExchanges` hook (renamed from `EthExchnges` and `useEthExchanges`)
- Restructured exchange data as a TypeScript object export
- Components associated with `/get-eth` page were also reviewed and updated to latest site standards

## Todo—Migrate/update remaining associated components:
- [x] `EthPriceCard`
- [x] `ButtonLink`
- [x] `FeedbackCard`
- [x] `InfoBanner`

## Related Issue
https://www.notion.so/efdn/Migrate-React-pages-8acee0dc80d04e55b109e9852c20ef2f?pvs=4

## Preview link
https://deploy-preview-152--ethereum-org-fork.netlify.app/en/get-eth
https://deploy-preview-152--ethereum-org-fork.netlify.app/tr/get-eth
https://deploy-preview-152--ethereum-org-fork.netlify.app/zh/get-eth

wackerow committed:
instantiate /get-eth page

wackerow committed:
implement CentralizedExchanges and associated hook

wackerow committed:
Delete exchanges-by-country.csv

wackerow committed:
update associated get-eth page components

wackerow committed:
extract resizeImage util

wackerow committed:
add lastDeployDate to getStaticProps

wackerow committed:
add LoadingState<T> type

wackerow committed:
migrate EthPriceCard

wackerow committed:
Merge branch 'dev' into get-eth-page

wackerow commented on src/hooks/useCentralizedExchanges.ts:
```diff
@@ -0,0 +1,360 @@
+  const lastUpdated = getLocaleTimestamp(
+    locale as Lang,
+    // TODO: Set up git log fetch; confirm no SSR issues
+    // data.timestamp.parent.fields.gitLogLatestDate
```
Another TODO worth noting... need to get the last time the exchange list was updated. We were using `git log` before; using `getStaticProps` I believe we should still be able to do this

- [x] Set up `git log` fetch

wackerow committed:
Merge branch 'dev' into get-eth-page

wackerow committed:
ilnting—clean up prop types and import ordering

wackerow commented on src/components/CentralizedExchanges/index.tsx:
```diff
@@ -97,32 +92,32 @@ const EthExchanges = () => {
-        options={exchangesByCountry}
-        onChange={handleSelectChange}
+        options={selectOptions}
+        onChange={handleSelectChange as any} // TODO: Fix typing
```
@pettinarip Ran into type errors here... the prop type for `handleSelectChange` was altered slightly and now it's complaining about the type not matching for `onChange`... it's working as is, but would prefer to not have the use of `any` here

pettinarip replied:
Not sure if I understand why this had to be changed. Did we need to refactor this logic?

wackerow committed:
patch percentage decimal value

wackerow committed:
create getLastModifiedDateByPath using git log

wackerow committed:
getLastModifiedDateByPath for exchange data

wackerow committed:
update Button components

wackerow committed:
update PageHero and CallToAction

ts ignore button story

wackerow committed:
add matomo tracking to ButtonLink

wackerow committed:
update FeedbackCard, implement useSurvey + matomo

wackerow committed:
update InfoBanner component

wackerow committed:
Merge branch 'dev' into get-eth-page

wackerow committed:
update image handling

wackerow committed:
rm resizeImage from useCentralizedExchanges hook

wackerow committed:
patch get-eth hero image sizing

corwintines committed:
Merge branch 'dev' into get-eth-page

corwintines committed:
fix responsive image issue  in calloutbanner

corwintines committed:
small refactor

corwintines reviewed (approved):
Nice work @wackerow!

wackerow commented:
Awesome, thanks for that patch @corwintines! It's not pixel-for-pixel, but it at least resizes appropriately as the screen width scales without weird overflows.
Separately I'll go through and deprecate entirely the use of `resizeImage` in favor of approach described here: <https://nextjs.org/docs/pages/api-reference/components/image#responsive-images>

corwintines committed:
Merge branch 'dev' into get-eth-page

corwintines committed:
remove resizeImage

wackerow committed:
Delete resizeImage.ts

wackerow committed:
patch image sizing

use shared iconImageProps, using Next.js recommended approach: https://nextjs.org/docs/pages/api-reference/components/image

pettinarip commented on src/pages/bug-bounty.tsx:
```diff
@@ -418,7 +417,7 @@ const BugBountiesPage = () => {
       image: useColorModeValue(tekuDark, tekuLight),
     },
-  ].map((client) => ({ ...client, image: resizeImage(client.image, 24) }))
+  ].map((client) => ({ ...client, image: client.image }))
```
no need for the map, right?
```suggestion
  ]
```

pettinarip commented on src/pages/bug-bounty.tsx:
```diff
@@ -438,6 +437,14 @@ const BugBountiesPage = () => {
   ]
+
+  const iconImageProps = {
+    sizes: "60px",
```
Since these are going to be fixed width images, there is no need to specify the `sizes` prop.

pettinarip commented on src/pages/bug-bounty.tsx:
```diff
@@ -438,6 +437,14 @@ const BugBountiesPage = () => {
+    style: {
+      width: "60px",
+      height: "auto",
+    },
```
We need to pass these props to the `Image` component to tell Next to resize the images accordingly.

In this case, the `width` will be enough as this is a static import and Next already knows the aspect ratio of the image. With the aspect ratio it can automatically calculate the height.

https://nextjs.org/docs/pages/api-reference/components/image#sizes

pettinarip committed:
refactor image size for fixed images

pettinarip commented on src/pages/get-eth.tsx:
```diff
@@ -0,0 +1,482 @@
+      lastDataUpdateDate,
+    },
+  }
+}) satisfies GetStaticProps<SSRConfig>
```
`lastDeployDate` and `lastDataUpdateDate` are missing from the Props declaration. We are only passing `SSRConfig`.

```ts
type Props = SSRConfig & {
  lastDeployDate..
}
```

pettinarip commented on src/pages/get-eth.tsx:
```diff
@@ -0,0 +1,482 @@
+  }
+}) satisfies GetStaticProps<SSRConfig>
+
+const GetEthPage = ({ lastDataUpdateDate }) => {
```
We can infer the prop types with `InferGetStaticPropsType`

pettinarip commented on src/lib/utils/gh.ts:
Nice!

pettinarip commented on src/components/CardList.tsx:
```diff
@@ -43,18 +43,33 @@ const CardContainer = (props: StackProps) => (
+          src={image}
+          alt={alt ?? ""}
+          sizes="20px"
+          style={{ width: "20px", height: "auto" }}
```
Related to the previous comment about the `Image` component. For fixed and static imported images, we don't need to use the `sizes` prop and we need to tell the width we want to use.

```ts
// this should be enough
<Image src={image} width={20} />
```

pettinarip replied:
While testing this I noticed that our `Image` component uses not only the Next Image width but also the Chakra width, `width: var(--eth-sizes-20)` (20rem), and this is not what we want. As a workaround, we can use the Next Image component in the meantime.

pettinarip committed:
fix types in the get-eth page

pettinarip committed:
fix image width for fixed images

pettinarip commented on src/components/Buttons/ButtonLink.tsx:
```diff
@@ -1,14 +1,28 @@
+    eventAction: "clicked",
+    eventName: "ButtonLink clicked",
+    eventValue: (props.to ?? props.href) as string,
+  }
```
Why are we adding this? do we want to keep track of every `ButtonLink` click? I'd lean towards removing this or discuss it with the team before we implement it.

Also, this is out of scope, we are changing behavior of the component.

wackerow replied:
Notion card: [Add customEventOptions prop back to ButtonLink & Link](https://www.notion.so/efdn/Add-customEventOptions-prop-back-to-ButtonLink-Link-ef2e0044450b4399b681fce5b3ce937d?pvs=4)

Was related to this card

wackerow replied:
This does add a default though; happy to remove this and just have an optional prop that can be passed.

pettinarip replied:
Got it, thanks.

We shouldn't need this in `ButtonLink`, as this action should be handled by the underlying `Link` component. The Link component already handles the matomo events, right? that is how we were doing it in the main repo.

pettinarip reviewed (changes requested):
Good job @wackerow. Overall it looks good on the preview deploy.

I'll push a few commits to fix things related to the use of images & types. Left some comments about it.

On the other hand, I see a lot of tiny refactors and, in some cases, some of them changing the styles and behavior. I'm not against them but lets try to keep them separated to be more organized, minimize potential bugs, and minimize the QA work for the migration.

wackerow committed:
remove fallback matomo event on ButtonLink

wackerow commented:
@pettinarip 
> Not sure if I understand why this had to be changed. Did we need to refactor this logic?

The data structure changed a little for the exchanges when converting from `csv` to a `ts` object...

```ts
  exchanges: { [key in ExchangeName]: StringBoolean }
  // became
  exchanges: string[]
```

Just like before we're using the same type for the select options as we are in the change handler.

I tried reverting some of the naming adjustments and state simplification but that didn't seem to change anything. Could refactor the JSON structure to more closely match the `"TRUE"` and `"FALSE"` usage we had before but that seemed kinda silly when not importing from CSV

wackerow committed:
Merge branch 'dev' into get-eth-page

pettinarip commented:
> @pettinarip
> 
> > Not sure if I understand why this had to be changed. Did we need to refactor this logic?
> 
> The data structure changed a little for the exchanges when converting from `csv` to a `ts` object...
> 
> ```ts
>   exchanges: { [key in ExchangeName]: StringBoolean }
>   // became
>   exchanges: string[]
> ```
> 
> Just like before we're using the same type for the select options as we are in the change handler.
> 
> I tried reverting some of the naming adjustments and state simplification but that didn't seem to change anything. Could refactor the JSON structure to more closely match the `"TRUE"` and `"FALSE"` usage we had before but that seemed kinda silly when not importing from CSV

Ahh got it 👍🏼 I'm fine with the current structure. I don't think it is necessary to change it now.

corwintines committed:
Merge branch 'dev' into get-eth-page
